File: 1036-rotting-oranges/solution.cpp
class Solution {
private:
    int M, N;
    
    bool check(vector<vector<int>>& g, queue<vector<int>>& q, int i, int j) {
        if (i > -1 && i < M && j > -1 && j < N && g[i][j] == 1) {
            g[i][j] = 2;
            q.push({i, j});
            return true;
        }
        return false;
    }

    int bfs(vector<vector<int>>& g, queue<vector<int>>& q, int c) {
        int t = 0;
        while(!q.empty()) {
            int s = q.size();
            for(int k = 0; k < s; k++) {
                vector<int> a = q.front(); q.pop();
                int i = a[0], j = a[1];
                if (check(g, q, i-1, j)) c--;
                if (check(g, q, i+1, j)) c--;
                if (check(g, q, i, j-1)) c--;
                if (check(g, q, i, j+1)) c--;
            }
            t++;
        }
        return c != 0 ? -1 : max(0, t-1);
    }

public:
    int orangesRotting(vector<vector<int>>& g) {
        M = g.size(), N = g[0].size();
        int ans = -1, c = 0;
        queue<vector<int>> q;

        for(int i = 0; i < M; i++) {
            for(int j = 0; j < N; j++) {
                if (g[i][j] == 1) c++;
                else if (g[i][j] == 2) q.push({i,j});
            }
        }

        return bfs(g, q, c);
    }
};

File: 1036-rotting-oranges/solution.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])
        queue, fresh, time = deque(), 0, 0

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 1: fresh += 1
                elif grid[r][c] == 2: queue.append((r, c))

        while fresh and queue:
            for i in range(len(queue)):
                r, c = queue.popleft()
                for dr, dc in [[0, 1], [0, -1], [1, 0], [-1, 0]]:
                    row, col = r + dr, c + dc
                    if (
                        row in range(rows) and
                        col in range(cols) and
                        grid[row][col] == 1
                    ):
                        grid[row][col] = 2
                        queue.append((row, col))
                        fresh -= 1
            time += 1

        return -1 if fresh else time

